store/src/java/com/zimbra/cs/service/formatter/LdifFormatter.java

```diff
@@ -0,0 +1,125 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import javax.mail.MessagingException;
+import javax.mail.Part;
+import javax.servlet.ServletException;
+
+import com.zimbra.common.localconfig.LC;
+import com.zimbra.common.mime.MimeConstants;
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.common.util.HttpUtil;
+import com.zimbra.cs.mailbox.MailItem;
+import com.zimbra.cs.service.UserServlet;
+import com.zimbra.cs.service.UserServletContext;
+import com.zimbra.cs.service.UserServletException;
+import com.zimbra.cs.service.formatter.FormatterFactory.FormatType;
+import com.zimbra.cs.service.util.UserServletUtil;
+import com.zimbra.cs.servlet.util.JettyUtil;
+
+public class LdifFormatter extends Formatter {
+
+    private static final String NEW_LINE = "\r\n";
+
+    @Override
+    public FormatType getType() {
+        return FormatType.LDIF;
+    }
+
+    @Override
+    public void formatCallback(UserServletContext context) throws UserServletException,
+        ServiceException, IOException, ServletException, MessagingException {
+        // Disable the jetty timeout
+        disableJettyTimeout(context);
+        Iterator<? extends MailItem> contacts = null;
+        StringBuilder sb = new StringBuilder();
+        try {
+            contacts = getMailItems(context, -1, -1, Integer.MAX_VALUE);
+            ArrayList<Map<String, String>> allContacts = new ArrayList<Map<String, String>>();
+            HashSet<String> fields = new HashSet<String>();
+            UserServletUtil.populateContactFields(contacts, context.targetMailbox, context.opContext, allContacts, fields);
+            ArrayList<String> allFields = new ArrayList<String>();
+            allFields.addAll(fields);
+            Collections.sort(allFields);
+            for (Map<String, String> contactMap : allContacts) {
+                toLDIFContact(allFields, contactMap, sb);
+                sb.append(NEW_LINE);
+            }
+        } finally {
+            if (contacts instanceof QueryResultIterator)
+                ((QueryResultIterator) contacts).finished();
+        }
+        String filename = context.itemPath;
+        if (filename == null || filename.length() == 0)
+            filename = "contacts";
+        if (filename.toLowerCase().endsWith(".ldif") == false) {
+            filename = filename + ".ldif";
+        }
+        String cd = HttpUtil.createContentDisposition(context.req, Part.ATTACHMENT, filename);
+        context.resp.addHeader("Content-Disposition", cd);
+        context.resp.setCharacterEncoding(context.getCharset().name());
+        context.resp.setContentType(MimeConstants.CT_TEXT_LDIF);
+        context.resp.getWriter().print(sb.toString());
+    }
+
+    private void toLDIFContact(List<String> fields, Map<String, String> contact, StringBuilder sb) {
```

Comment:
ref: https://stackoverflow.com/questions/33433453/conversion-of-csv-file-to-ldif-file-using-java/45622057#45622057

The solution offered seems incomplete when it comes to dealing with any possible data that may show up. If you look at [RFC2849](https://tools.ietf.org/html/rfc2849) you'll find recommendations like (both on Page 5):

```
 4)  ... Any
     value that contains characters other than those defined as
     "SAFE-CHAR", or begins with a character other than those
     defined as "SAFE-INIT-CHAR", above, MUST be base-64 encoded.
     Other values MAY be base-64 encoded.
```
where SAFE-INIT-CHAR is defined as

```
SAFE-INIT-CHAR       = %x01-09 / %x0B-0C / %x0E-1F /
                       %x21-39 / %x3B / %x3D-7F
                       ; any value <= 127 except NUL, LF, CR,
                       ; SPACE, colon (":", ASCII 58 decimal)
                       ; and less-than ("<" , ASCII 60 decimal)
```
and

```
 8)  Values or distinguished names that end with SPACE SHOULD be
     base-64 encoded.
```

Unless I'm missing something, the code should be updated to handle encoding properly if you want a more complete and robust solution.